File: modules/test-e2e-angular/src/main/java/europa/edit/pages/ProposalViewerPage.java
package europa.edit.pages;

import europa.edit.util.Constants;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.CacheLookup;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.PageFactory;
import org.openqa.selenium.support.pagefactory.AjaxElementLocatorFactory;

import static europa.edit.util.E2eUtil.*;

public class ProposalViewerPage extends HeaderPage {

    private final WebDriver driver;

    @FindBy(xpath = "//*[@class='eui-tab-item__label' and text()='Milestones']")
    @CacheLookup
    WebElement MILESTONES_TAB;

/*
    @FindBy(xpath = "//*[@class='eui-tab-item__label' and text()='Collaborators']")
    @CacheLookup
    WebElement COLLABORATORS_TAB;
*/

    @FindBy(xpath = "//*[@class='eui-label' and text()='Actions']")
    @CacheLookup
    WebElement ACTIONS_BTN;

    @FindBy(css = ".eui-u-flex-align-items-start h1")
    @CacheLookup
    WebElement PROPOSAL_TITLE;

/*    @FindBy(css = "app-proposal-details .eui-fieldset__container")
    @CacheLookup
    WebElement PROPOSAL_DETAILS_CONTAINER;*/

    @FindBy(css = "div.eui-tab-item--active .eui-tab-item__label")
    @CacheLookup
    WebElement ACTIVE_TAB;

    @FindBy(css = "app-proposal-header header>button[translate='global.actions.close']")
    @CacheLookup
    WebElement CLOSE_BTN;

    @FindBy(xpath = "//*[text()='Cover Page']")
    @CacheLookup
    WebElement COVER_PAGE;

    @FindBy(xpath = "//*[text()='Explanatory Memorandum']")
    @CacheLookup
    WebElement EXP_MEMO;

    @FindBy(xpath = "//*[text()='Legal Act']")
    @CacheLookup
    WebElement LEGAL_ACT;

    @FindBy(xpath = "//*[text()='Financial Statement']")
    @CacheLookup
    WebElement FINANCE_STATEMENT;

    @FindBy(xpath = "//*[text()='Annexes']")
    @CacheLookup
    WebElement ANNEXES;

    @FindBy(xpath = "//*[text()='Annexes']//ancestor::div[@class='flex-grow-1']//parent::div//button[1]")
    @CacheLookup
    WebElement ANNEX_ADD_BTN;

/*    @FindBy(xpath = "//*[text()='Annexes']//ancestor::div[@class='flex-grow-1']//parent::div//button[2]")
    @CacheLookup
    WebElement ANNEX_REORDER_BTN;*/

    By deleteButton = By.cssSelector("button.eui-list-item--danger");
    By downloadButton = By.xpath("//button//*[contains(text(),'Download')]");
    By changeTitleButton = By.xpath("//button//*[contains(text(),'Change title')]");
    By annexFound = By.xpath("//*[contains(text(),'annexes found')]");
    String tab = "div.eui-tab-item:nth-child(%d)";
    String annexTitle = "//table//tbody//tr[%d]//td//a";

    public ProposalViewerPage(WebDriver driver) {
        super(driver);
        this.driver = driver;
        waitForLoadingProgressBarToDisappear(driver);
        AjaxElementLocatorFactory ajax = new AjaxElementLocatorFactory(driver, Constants.TIMEOUT_DELAY);
        PageFactory.initElements(ajax, this);
    }

    public String getProposalTitle() {
        return PROPOSAL_TITLE.getText();
    }

    public boolean isCloseButtonDisplayed() {
        return CLOSE_BTN.isDisplayed();
    }

    public boolean isCloseButtonEnabled() {
        return CLOSE_BTN.isEnabled();
    }

    public void clickOnActionButton() {
        elementClick(driver, ACTIONS_BTN);
    }

    public void clickDownloadButton() {
        elementClick(driver, driver.findElement(downloadButton));
    }

    public void clickOnDeleteButton() {
        driver.findElement(deleteButton).click();
    }

    public void clickOnCloseButton() {
        elementClick(driver, CLOSE_BTN);
    }

    public String getTabName(int tabIndex) {
        return getElementText(driver.findElement(By.cssSelector(String.format(tab, tabIndex))));
    }

    public String getActiveTabName() {
        return getElementText(ACTIVE_TAB);
    }

    public int getNumberOfAnnexes() {
        return Integer.parseInt(String.valueOf(getElementText(driver.findElement(annexFound)).charAt(0)));
    }

    public String getAnnexTitle(int annexIndex) {
        return getElementText(driver.findElement(By.xpath(String.format(annexTitle, annexIndex))));
    }

    public boolean isCoverPageSectionDisplayed() {
        return waitForElementTobeDisPlayed(driver, COVER_PAGE);
    }

    public boolean isExpMemoSectionDisplayed() {
        return waitForElementTobeDisPlayed(driver, EXP_MEMO);
    }

    public boolean isLegalActSectionDisplayed() {
        return waitForElementTobeDisPlayed(driver, LEGAL_ACT);
    }

    public boolean isFinanceStatementSectionDisplayed() {
        return waitForElementTobeDisPlayed(driver, FINANCE_STATEMENT);
    }

    public boolean isAnnexesSectionDisplayed() {
        return waitForElementTobeDisPlayed(driver, ANNEXES);
    }

    public boolean contentDisplayedUnderFinancialStatement(String content) {
        return waitForElementTobeDisPlayed(driver, getWebElementByText(content));
    }

    public boolean contentDisplayedUnderFAnnexesSection(String content) {
        return waitForElementTobeDisPlayed(driver, getWebElementByText(content));
    }

    public WebElement getWebElementByText(String text) {
        return driver.findElement(By.xpath("//*[contains(text(),'" + text + "')]"));
    }

    public void clickAnnexAddButton() {
        elementClick(driver, ANNEX_ADD_BTN);
        waitForLoadingProgressBarToDisappear(driver);
    }

    public void clickChangeTitleButton() {
        elementClick(driver, driver.findElement(changeTitleButton));
    }

    public void clickMileStonesTab() {
        MILESTONES_TAB.click();
    }

    public void clickOnCoverPageLInk() {
        COVER_PAGE.click();
    }
}

File: modules/test-cypress/cypress/e2e/pages/proposalViewerPage.js
import headerPage from './headerPage';

class proposalViewerPage extends headerPage{
    elements = {
        proposalTitle: () => cy.get('app-proposal-header h1'),
        favouriteIconBtn: () => cy.get('eui-icon.eui-icon--size-l'),
        actionBtn: () => cy.get('app-proposal-actions-dropdown button'),
        downloadBtn:() => cy.get('.eui-dropdown-content button').contains('Download'),
        deleteBtn: () => cy.get('button.eui-list-item--danger'),
        closeBtn: () => cy.contains("Close"),
        legalActLink: () => cy.get('.eui-tab-content-wrapper eui-card').contains('Legal Act'),
        coverPageLink: () => cy.get('.eui-tab-content-wrapper eui-card').contains('Cover Page'),
        expMemoLink: () => cy.get('.eui-tab-content-wrapper eui-card').contains('Explanatory Memorandum'),
        annexesSection: () => cy.get('.eui-tab-content-wrapper eui-card').contains('Annexes'),
        AddBtnfinancialStatement: () => cy.xpath("//div/*[text()='Financial Statement']//ancestor::div[contains(@class,'eui-u-flex')]//button//span[text()=' Add ']"),
        financialStatementLink: () => cy.xpath("//a/*[text()='Financial Statement']"),
        deleteBtnfinancialStatement: () => cy.xpath("//a/*[text()='Financial Statement']//ancestor::div[contains(@class,'eui-u-flex')]//button//span[text()=' Delete ']"),
        activeTab: () => cy.get('div.eui-tab-item--active'),
        draftsTab: () => cy.get('div.eui-tabs__items-wrapper .eui-tab-item').contains('Drafts'),
        milestoneTab: () => cy.get('div.eui-tabs__items-wrapper .eui-tab-item').contains('Milestones'),
        collaboratorsTab: () => cy.get('div.eui-tabs__items-wrapper .eui-tab-item').contains('Collaborators'),
        proposalDetailsTab: () => cy.get('div.eui-tabs__items-wrapper .eui-tab-item').contains('Details'),
        AddBtnAnnex: () => cy.xpath("//div/*[text()='Annexes']//ancestor::div[contains(@class,'eui-u-flex')]//button//span[text()=' Add ']"),
        reOrderBtnAnnex: () => cy.xpath("//div/*[text()='Annexes']//ancestor::div[contains(@class,'eui-u-flex')]//button//span[text()=' Reorder ']"),
        noAnnexPresent: () => cy.contains("There is no annex to this document"),
        annexCount: () => cy.get('eui-card .eui-card-content table tbody tr'),
        changeTitleBtn: () => cy.get("button[translate$='actions.dropdown.edit-title']"),
        deleteBtnFromActionMenu: () => cy.get("button[translate='global.actions.delete']")
    }

    clickCloseBtn(){
        this.elements.closeBtn().click();
    }

    clickLegalActLink(){
        this.elements.legalActLink().click();
    }

    clickOnActionButton(){
        this.elements.actionBtn().click();
    }

    clickDownloadButton(){
        this.elements.downloadBtn().click();
        cy.wait(5000);
    }

    clickOnDeleteButton(){
        this.elements.deleteBtn().click();
    }

    clickCoverPageLink(){
        this.elements.coverPageLink().click();
    }

    clickAddAnnexBtn(){
        this.elements.AddBtnAnnex().click();
    }

    clickNthAnnex(annexNumber){
        this.elements.annexCount().eq(annexNumber-1).find('a').click();
    }

    clickActionsMenuOfAnnex(annexNumber){
        this.elements.annexCount().eq(annexNumber-1).find("button[aria-label='Annex Actions'] .eui-icon-more-vertical").trigger('mouseover').click();
    }

    clickChangeTitleBtn(){
        this.elements.changeTitleBtn().click();
    }

    getTitleElementOfAnnex(annexNumber){
        return this.elements.annexCount().eq(annexNumber-1).find('a strong');
    }
    
    clickDeleteBtnFromActionMenu(){
        this.elements.deleteBtnFromActionMenu().click();
    }

    clickMilestonesTab(){
        this.elements.milestoneTab().click();
    }
}
export default new proposalViewerPage();

import '@cypress/xpath';